components/builder-core/src/http_client.rs

```diff
@@ -0,0 +1,98 @@
+use std::collections::HashMap;
+
+use reqwest::{header::HeaderMap,
+              Body,
+              Client,
+              Method,
+              Proxy,
+              Response};
+
+use url::Url;
+
+use crate::error::{Error,
+                   Result};
+
+pub const BLDR_USER_AGENT: &str = "Habitat-Builder";
+pub const APPLICATION_JSON: &str = "application/json";
+pub const FORM_URL_ENCODED: &str = "application/x-www-form-urlencoded";
+
+pub enum Auth<'a> {
+    None,
+    Basic(&'a str, &'a str),
+    Bearer(&'a str),
+}
+
+pub struct HttpCall<'a> {
+    pub method:  Method,
+    pub url:     &'a str,
+    pub headers: Option<HeaderMap>,
+    pub params:  Option<&'a HashMap<&'a str, &'a str>>,
+    pub body:    Option<Body>,
+    pub auth:    Auth<'a>,
+}
+
+#[derive(Clone)]
+pub struct HttpClient(Client);
+
+impl HttpClient {
+    pub fn new(url: &str, headers: HeaderMap) -> Self {
+        let mut client = Client::builder();
+
+        trace!("HttpClient: checking proxy for url: {:?}", url);
+        let url = Url::parse(url).expect("valid client url must be configured");
+
+        if let Some(proxy_url) = env_proxy::for_url(&url).to_string() {
+            if url.scheme() == "http" {
+                trace!("Setting http_proxy to {}", proxy_url);
+                match Proxy::http(&proxy_url) {
+                    Ok(p) => {
+                        client = client.proxy(p);
+                    }
+                    Err(e) => warn!("Invalid proxy, err: {:?}", e),
+                }
+            }
+
+            if url.scheme() == "https" {
+                trace!("Setting https proxy to {}", proxy_url);
+                match Proxy::https(&proxy_url) {
+                    Ok(p) => {
+                        client = client.proxy(p);
+                    }
+                    Err(e) => warn!("Invalid proxy, err: {:?}", e),
+                }
+            }
+        } else {
+            trace!("No proxy configured for url: {:?}", url);
+        }
+
+        HttpClient(client.default_headers(headers).build().unwrap())
+    }
+
+    pub fn http_call<'a>(&self, spec: HttpCall<'a>) -> Result<Response> {
```

Comment:
I think this method (and the `HttpCall` struct as well, since it seems to be the only user) can be eliminated in favor of exposing the [`reqwest::RequestBuilder`](https://docs.rs/reqwest/0.9.18/reqwest/struct.RequestBuilder.html) interface.

If we add
```rust
impl std::ops::Deref for HttpClient {
    type Target = Client;

    fn deref(&self) -> &Self::Target { &self.0 }
}
```
and some error-converting functions, we can make calls like
```rust
let resp = match self.inner
                     .http_call(HttpCall { method:  Method::PUT,
                                           url:     &url,
                                           headers: None,
                                           params:  None,
                                           body:    Some(file.into()),
                                           auth:    Auth::None, })
                     .map_err(ArtifactoryError::BuilderCore)
{
```
into
```rust
let resp = match self.inner.put(&url).body(file).send()
```
and calls like:
```rust
let auth = if token.is_some() {
    Auth::Bearer(token.unwrap())
} else {
    Auth::None
};

let mut resp = self.inner.http_call(HttpCall { method: Method::GET,
                                                url: &url_path,
                                                headers: None,
                                                params: Some(&query),
                                                body: None,
                                                auth })?;
```
into
```rust
let mut request = self.inner.get(&url_path).query(&query);

if let Some(token) = token {
    request = request.bearer_auth(token);
}

let mut resp = request.send()?;
```

I tried going through a number of the callers to see if it didn't work well, but so far, I haven't found one. For the sake of avoiding duplicate work, [here's a diff](https://github.com/habitat-sh/builder/files/3341174/builder-pr-1109-jbauman.diff.zip) of what I did so far.

Reply:
This is a really good observation. Ideally we could encapsulate out needing to expose reqwest at all, but that's a lot more work, and for now going this route seems like a much simpler approach. Will make the suggested change.